File: leetcode/python/46. 全排列.py
#!/bin/bash python
"""
46. 全排列

思路：
DFS
"""

class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        """
        回溯法
        因为没有重复，所以不需要进行排序
        """
        # 特殊情况
        if nums == []:
            return []
        
        def dfs(nums, ret, i):
            """
            i表示正在进行第i索引
            """
            if i >= len(nums):
                ret.append(nums[:])
                return
            for j in range(i, len(nums)):
                nums[i], nums[j] = nums[j], nums[i]
                dfs(nums, ret, i + 1)
                nums[i], nums[j] = nums[j], nums[i]
            return

        # 正式调用
        ret = []
        dfs(nums, ret, 0)
        return ret

File: leetcode/cpp/46. 全排列.cpp
class Solution {
public:
    void swap(vector<int>& vec, int a, int b) {
        int tmp = vec[a];
        vec[a] = vec[b];
        vec[b] = tmp;
    }

    void dfs(vector<vector<int>>& ret, vector<int>& nums, int i) {
        if (i >= nums.size()) {
            ret.push_back(nums);
            return;
        }
        for (int j = i; j < nums.size(); j++) {
            swap(nums, i, j);
            dfs(ret, nums, i + 1);
            swap(nums, i, j);
        }
        return;
    }

    vector<vector<int>> permute(vector<int>& nums) {
        vector<vector<int>> ret;
        if (!nums.empty()) {
            dfs(ret, nums, 0);
        }
        return ret;
    }
};